### PR TITLE
Add flat_index_at utility function to record_type

### PR DESCRIPTION
**Pull request by Neverlord**
This PR adds a feature needed by #191.

### CODE REVIEW

File: libvast/src/type.cpp [hunk 1]
```diff
@@ -381,6 +381,39 @@ const type* record_type::at(const offset& o) const {
...
 }
 
+caf::optional<size_t> record_type::flat_index_at(offset o) const {
+  // Empty offsets are invalid.
+  if (o.empty())
+    return caf::none;
+  // Bounds check.
+  if (o[0] >= fields.size())
+    return caf::none;
+  // o = [1] picks the second element. However, we still need the total
```
Comment (mavam) [inline libvast/src/type.cpp]: What do you mean with `o = [1]`?

File: libvast/src/type.cpp [hunk 2]
```diff
@@ -381,6 +381,39 @@ const type* record_type::at(const offset& o) const {
...
+  // amount of nested elements of the first element.
+  size_t flat_index = 0;
+  for (size_t i = 0; i < o[0]; ++i)
+    flat_index += flat_size(fields[i].type);
+  // Now, we know how many fields are on the left. We're done if our the offset
+  // points a non-record field in this record.
+  auto& field = fields[o[0]];
+  auto field_record = get_if<record_type>(field.type);
```
Comment (mavam) [inline libvast/src/type.cpp]: Nit: I would find`record_field` more intuitive here.

File: libvast/src/type.cpp [hunk 3]
```diff
@@ -381,6 +381,39 @@ const type* record_type::at(const offset& o) const {
...
+  if (o[0] >= fields.size())
+    return caf::none;
+  // o = [1] picks the second element. However, we still need the total
+  // amount of nested elements of the first element.
+  size_t flat_index = 0;
+  for (size_t i = 0; i < o[0]; ++i)
+    flat_index += flat_size(fields[i].type);
+  // Now, we know how many fields are on the left. We're done if our the offset
+  // points a non-record field in this record.
+  auto& field = fields[o[0]];
```
Comment (mavam) [inline libvast/src/type.cpp]: Do we need this aliased variable? Why not put it in the statement below?

File: libvast/src/type.cpp [hunk 4]
```diff
@@ -381,6 +381,39 @@ const type* record_type::at(const offset& o) const {
...
+  auto field_record = get_if<record_type>(field.type);
+  if (o.size() == 1) {
+    // Sanity check: the offset is invalid if points to a record type.
```
Comment (mavam) [inline libvast/src/type.cpp]: `...if [it] points..`

File: libvast/test/type.cpp [hunk 5]
```diff
@@ -215,6 +224,40 @@ TEST(record flattening/unflattening) {
...
+  CHECK_EQUAL(x.flat_index_at(make_offset({0, 0, 0})), os(0u));
+  CHECK_EQUAL(x.flat_index_at(make_offset({0, 0, 1})), os(1u));
+  CHECK_EQUAL(x.flat_index_at(make_offset({0, 1, 0, 0})), os(2u));
+  CHECK_EQUAL(x.flat_index_at(make_offset({0, 1, 1})), os(3u));
+  CHECK_EQUAL(x.flat_index_at(make_offset({0, 2})), os(4u));
+  CHECK_EQUAL(x.flat_index_at(make_offset({1, 0})), os(5u));
+  CHECK_EQUAL(x.flat_index_at(make_offset({0})), invalid);
+  CHECK_EQUAL(x.flat_index_at(make_offset({0, 0})), invalid);
+  CHECK_EQUAL(x.flat_index_at(make_offset({1})), invalid);
+  CHECK_EQUAL(x.flat_index_at(make_offset({2})), invalid);
```
Comment (mavam) [inline libvast/test/type.cpp]: Nice collection of tests!

File: libvast/src/type.cpp (hunk 1 above)
Comment (Neverlord) [reply libvast/src/type.cpp]: Just an example.

**Pull request merged**